test/unit/iterator_type_test.lua

```diff
@@ -0,0 +1,536 @@
+                {2, "2"},
+                {3, "1"}
+            })
+        end)
+        task:kill()
+    end
+end
+
+function g.test_hash_index_all()
+    -- without start key
+    helpers.iteration_result = {}
+    g.hash:insert({3, "1"})
+    g.hash:insert({2, "2"})
+    g.hash:insert({1, "3"})
+
+    local task = expirationd.start("clean_all", g.hash.id, helpers.is_expired_debug,
+            {iterator_type = "ALL"})
+    -- wait for tuples expired
+    helpers.retrying({}, function()
+        t.assert_equals(helpers.iteration_result, {
+            {1, "3"},
+            {2, "2"},
+            {3, "1"}
+        })
+    end)
+    task:kill()
+
+    -- with start_key
+    -- Implicit behavior hash index ALL with start key
+    helpers.iteration_result = {}
+    g.hash:insert({1, "3"})
+    g.hash:insert({2, "2"})
+    g.hash:insert({3, "1"})
+
+    task = expirationd.start("clean_all", g.hash.id, helpers.is_expired_debug,
+            {iterator_type = "ALL", start_key = 2})
+    -- wait for tuples expired
+    helpers.retrying({}, function()
+        t.assert_equals(helpers.iteration_result, {
+            {3, "1"},
+            {2, "2"},
+            {1, "3"}
+        })
+    end)
+    task:kill()
+end
+
+function g.test_tree_index_eq()
+    for _, space in pairs({g.tree, g.vinyl}) do
+        -- without start key
+        helpers.iteration_result = {}
+        space:insert({1, "3"})
+        space:insert({2, "2"})
+        space:insert({3, "1"})
+
+        local task = expirationd.start("clean_all", space.id, helpers.is_expired_debug,
+                {iterator_type = "EQ"})
+        -- wait for tuples expired
+        helpers.retrying({}, function()
+            t.assert_equals(helpers.iteration_result, {
+                {1, "3"},
+                {2, "2"},
+                {3, "1"}
+            })
+        end)
+        task:kill()
+
+        -- with start_key
+        helpers.iteration_result = {}
+        space:insert({1, "1", nil, nil, 1})
+        space:insert({2, "2", nil, nil, 2})
+        space:insert({3, "3", nil, nil, 2})
+        space:insert({4, "4", nil, nil, 4})
+
+        task = expirationd.start("clean_all", space.id, helpers.is_expired_debug,
+                {iterator_type = "EQ", index = "non_unique_index", start_key = 2})
+        -- wait for tuples expired
+        helpers.retrying({}, function()
+            t.assert_equals(helpers.iteration_result, {
+                {2, "2", nil, nil, 2},
+                {3, "3", nil, nil, 2}
+            })
+        end)
+        task:kill()
+    end
+end
+
+function g.test_hash_index_eq()
+    -- without start key
+    --- throw:
+    --- main/119/worker of "clean_all" index.cc:127 E>
+    --- ER_PARTIAL_KEY: HASH index  does not support selects via a partial key (expected 1 parts, got 0).
+    --- Please Consider changing index type to TREE
+    --- Unclear error, select(nil, {iterator = "EQ"}) works with tree index and return all values, hash does not
```

Comment:
It's not unclear. Just HASH could find with "EQ" iterator only with full specified key. And such case looks really strange - "EQ" for HASH means I want to run my task to clear only single value (because hash can't non-unique)

I think such error should be thrown at validation stage.

Reply:
I wrote this to the fact that tarantool allows you to do this and gives a strange error, although as you wrote, it has to run through only one key. It seems to me that this needs to be fixed on the tarantool side, or removed from the documentation that you can use the EQ iterator for the hash index. Of course, if there are reasons not to change this in Tarantool, then I will validate this type of iterator

Reply:
```
box.space.test:create_index('hash', {type = 'hash'})
tarantool> box.space.test.index.hash:pairs(nil, {iterator = 'EQ'})
---
- error: HASH index  does not support selects via a partial key (expected 1 parts,
    got 0). Please Consider changing index type to TREE.
...
```

It should fail on validation stage.

Reply:
but there is no partial key here, or I'm wrong?

Reply:
`select(nil)` is the same as `select({})`. It's partial key.

Reply:
Let's add a validation for `nil` index value.

Reply:
> Let's add validation for `nil` index value.

It doesn't work so. You also should ban `{}` and `{box.NULL}`. It's bad way. Just check with `box.space.test.index.hash:pairs(nil, {iterator = 'EQ'})`.

Reply:
There was such a check, I just finished the test to check this validation